File: Algorithms/Average of Levels in Binary Tree/average-of-levels-in-binary-tree.js
// Source : https://leetcode.com/problems/average-of-levels-in-binary-tree/description/
// Author : Fish.Yu
// Date   : 2017-12-26

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function(root) {
  root.level = 0
  let q = [root]
  let ans = []
  let [sum, num, curLevel] = [0, 0, 0]

  while (q.length) {
    let item = q.shift()
    if (item.level === curLevel) {
      sum += item.val 
      num += 1
    } else {
      ans.push(sum / num)
      curLevel += 1
      sum = item.val
      num = 1
    }

    if (item.left) {
      item.left.level = item.level + 1
      q.push(item.left)
    }
    if (item.right) {
      item.right.level = item.level + 1
      q.push(item.right)
    }
  }

  ans.push(sum / num)
  return ans
};

File: Algorithms/Average of Levels in Binary Tree/average-of-levels-in-binary-tree.py
# Source : https://leetcode.com/problems/average-of-levels-in-binary-tree/description/
# Author : Fish.Yu
# Date   : 2017-12-26

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        root.level = 0
        q = [root]
        ans = []
        sum, num, curLevel = 0, 0, 0

        while len(q):
            item = q[0]
            del q[0]
            if item.level == curLevel:
                sum += item.val 
                num += 1
            else:
                ans.append(sum / num)
                curLevel += 1
                sum = item.val 
                num = 1
            
            if item.left:
                item.left.level = item.level + 1
                q.append(item.left)
            if item.right:
                item.right.level = item.level + 1
                q.append(item.right)
            
        ans.append(sum / num)
        return ans